File: mrhalls_maths/binomial_expansion.py
from math import comb

def binomial_expansion(expression, power):
    result = ""
    for k in range(power + 1):
        coefficient = comb(power, k)
        term_a = round(expression[0] ** (power - k), 2)
        term_b = round(expression[1] ** k, 2)
        term = round(coefficient * term_a * term_b, 2)
        result += f"{term} * {expression[2]}^{power - k} * {expression[3]}^{k} + "
    return result[:-3]  # Remove the extra "+ " at the end

# Test the function
expression = [1,2, "x", "y"]
power = 5
expansion = binomial_expansion(expression, power)
print(expansion)

File: maths/binomialexpansion.c
#include <stdio.h>
#include <math.h>

long long int binomialCoefficient(int n, int k) {
    if (k == 0 || k == n)
        return 1;
    else
        return binomialCoefficient(n - 1, k - 1) + binomialCoefficient(n - 1, k);
}

void binomialExpansion(int base_a, int base_b, int power) {
    for (int k = 0; k <= power; k++) {
        int coefficient = binomialCoefficient(power, k);
        double term_a = pow(base_a, power - k);
        double term_b = pow(base_b, k);
        double term = coefficient * term_a * term_b;
        printf("%.2f * x^%d * y^%d + ", term, power - k, k);
    }
}

int main() {
    int base_x = -2;
    int base_y = 1;
    int power = 5;
    binomialExpansion(base_x, base_y, power);
    
    return 0;
}
